src/lib/PerceptronPredictor.cc

```diff
@@ -143,6 +156,19 @@ void PerceptronPredictor::flush(uint64_t address) {
     }
     rasHistory_.erase(it);
   }
+
+  // If possible, pop instruction from FTQ
+  // if (!FTQ_.empty()) FTQ_.pop_back();
+  FTQ_.pop_back();
+
+  // Roll back global history
+  globalHistory_ >>= 1;
+}
+
+void PerceptronPredictor::addToFTQ(uint64_t address, bool taken) {
+  // Add instruction to the FTQ in event of reused prediction
+  FTQ_.emplace_back(taken, globalHistory_);
+  globalHistory_ = ((globalHistory_ << 1) | taken) & globalHistoryMask_;
```

Comment:
Could there ever be a scenario where `addToFTQ` and `predict` are called for the same branch instruction; adding said branch to the FTQ and globalHistory twice?

Reply:
See my reply to your comment re GenericPredictor.cc line 138